feat: add `callback` api to `WebSocketProxy`

WSH032 opened the pull request:
- bump `typing_extensions` from `4.5.0` to `4.12.0`
- explicitly add `anyio` as a dependency
- fix docs `host` typo
- fix broken `httpx` authentication docs link
- refactor `tool_4_test_fixture` to factory to support testing different ws apps

<!-- Thanks for contributing 💚
Given this is a project maintained by volunteers, please read this template to not waste your time, or ours! 😁 -->

# Summary

Close: #40

## example

### proxy server

```py
from contextlib import asynccontextmanager
from typing import AsyncIterator

from fastapi import FastAPI
from fastapi_proxy_lib.core.websocket import (
    CallbackPipeContextType,
    ReverseWebSocketProxy,
)
from httpx import AsyncClient
from starlette.websockets import WebSocket

proxy = ReverseWebSocketProxy(AsyncClient(), base_url="ws://echo.websocket.events/")


async def client_to_server_callback(pipe_context: CallbackPipeContextType[str]) -> None:
    with pipe_context as (sender, receiver):
        async for message in receiver:
            print(f"Received from client: {message}")
            await sender.send(f"CTS:{message}")
    print("client_to_server_callback end")


async def server_to_client_callback(
    pipe_context: CallbackPipeContextType[str],
) -> None:
    with pipe_context as (sender, receiver):
        async for message in receiver:
            print(f"Received from server: {message}")
            await sender.send("STC:{message}")
    print("server_to_client_callback end")


@asynccontextmanager
async def close_proxy_event(_: FastAPI) -> AsyncIterator[None]:
    """Close proxy."""
    yield
    await proxy.aclose()


app = FastAPI(lifespan=close_proxy_event)


@app.websocket("/{path:path}")
async def _(websocket: WebSocket, path: str = ""):
    return await proxy.proxy(
        websocket=websocket,
        path=path,
        client_to_server_callback=client_to_server_callback,
        server_to_client_callback=server_to_client_callback,
    )


# Then run shell: `uvicorn <your_py>:app --host 127.0.0.1 --port 8000`
# visit the app: `ws://127.0.0.1:8000/`
# you can establish websocket connection with `ws://echo.websocket.events`
```

### client

```python
from httpx_ws import aconnect_ws

async with aconnect_ws('ws://127.0.0.1:8000/') as ws:
    message = await ws.receive_text()
    print(f"Received: {message}")

    await ws.send_text('Hello, World!')

    message = await ws.receive_text()
    print(f"Received: {message}")
```

<!-- Write a small summary about what is happening here. -->

# Checklist

- [ ] I've read `CONTRIBUTING.md`.
- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.

WSH032 commented:
CC @IvesAwadi

It's almost done. Would you like to give it a try?

```bash
pip install git+https://github.com/WSH032/fastapi-proxy-lib.git@feat/ws-callback
```

This will pull the latest `fastapi-proxy-lib` from this PR.

---

Before merging, there are a few things that need to be addressed:

- Add tests
- Write documentation

And, I'm a bit unsatisfied with the API; is `client_to_server_callback` too long?

IvesAwadi commented:
> CC @IvesAwadi
> 
> It's almost done. Would you like to give it a try?
> 
> ```shell
> pip install git+https://github.com/WSH032/fastapi-proxy-lib.git@feat/ws-callback
> ```
> 
> This will pull the latest `fastapi-proxy-lib` from this PR.
> 
> Before merging, there are a few things that need to be addressed:
> 
> * Add tests
> * Write documentation
> 
> And, I'm a bit unsatisfied with the API; is `client_to_server_callback` too long?

It works perfectly, no issues on my end using Python 3.11.9. The `client_to_server_callback` I'd argue isn't long at all, the name make sense in this context.

WSH032 commented:
I just found a bug. The current implementation only supports a strict `one-receive-one-send` mode within a single loop. If this pattern is violated, such as `multiple receives and one send`, `one receive and multiple sends`, or `sending before receiving` within a single loop, it will result in a deadlock.

```py
async def callback(ctx: CallbackPipeContextType[str]) -> None:
    with ctx as (sender, receiver):
        # multiple receives and one send, dead lock!
        await receiver.receive()
        await receiver.receive()
        await sender.send("foo")

async def callback(ctx: CallbackPipeContextType[str]) -> None:
    with ctx as (sender, receiver):
        # one receive and multiple sends, dead lock!
        async for message in receiver:
            await sender.send("foo")
            await sender.send("bar")

async def callback(ctx: CallbackPipeContextType[str]) -> None:
    with ctx as (sender, receiver):
        # sending before receiving, dead lock!
        await sender.send("foo")
        async for message in receiver:
            await sender.send(message)
```

Unfortunately, we can't resolve this issue until the underlying logic is rewritten using `anyio` and `memory-object-streams`.

There is already a PR for `anyio`: #34. However, this PR still has many issues, and I currently don't have time to merge it.

WSH032 committed:
feat: add `callback` api to `WebSocketProxy`

- bump `typing_extensions` from `4.5.0` to `4.12.0`
- explicitly add `anyio` as a dependency
- fix docs `host` typo

WSH032 committed:
test: refactor `tool_4_test_fixture` to factory to support testing different ws apps

WSH032 committed:
test: add tests for WebSocketProxy callback

- test message modification using `callback` under normal conditions
- test WebSocket closing when `callback` encounters an exception

WSH032 committed:
feat: use `task.add_done_callback` to make sure callback-side pipe context exited

WSH032 committed:
docs: add docs for `WebSocketProxy` callback api

- fix broken `httpx` authentication docs link

WSH032 commented on docs/Usage/Advanced.md:
```diff
@@ -118,3 +119,19 @@ async def _(request: Request, path: str = ""):
 
 visit `/`, you will notice that the response body is printed to the console.
+
+## Modify WebSocket message
+
+In some cases, you might want to modify the content of the messages that the WebSocket proxy receives and sends to the client and target server.
+
+In version `0.2.0` of `fastapi-proxy-lib`, we introduced a [`callback API`][fastapi_proxy_lib.core.websocket.ReverseWebSocketProxy.proxy] for `WebSocketProxy` to allow you to do this.
+
+See example: [ReverseWebSocketProxy#with-callback][fastapi_proxy_lib.core.websocket.ReverseWebSocketProxy--with-callback]
+
+Also:
+
+- RFC: [#40](https://github.com/WSH032/fastapi-proxy-lib/issues/40)
+- PR: [#41](https://github.com/WSH032/fastapi-proxy-lib/pull/41)
+
+!!!example
+    The current implementation still has some defects. Read the [callback-implementation][fastapi_proxy_lib.core.websocket.BaseWebSocketProxy.send_request_to_target--callback-implementation] section, or you might accidentally shoot yourself in the foot.
```
@IvesAwadi What do you think of this description? Can you provide some specific use cases? This way, I can add them to the documentation.

IvesAwadi replied:
> @IvesAwadi What do you think of this description? Can you provide some specific use cases? This way, I can add them to the documentation.

The description is good and makes sense, took a bit to respond to this (had some issues with my email.)

WSH032 commented:
@IvesAwadi. This PR is completed. Do you think there are any areas that need improvement? If not, I will merge this PR tomorrow.

As a non-native English speaker, I’m unsure about the quality of the API documentation (including inline docstrings). I would greatly appreciate any suggestions for improvement.

If you want to preview the documentation in advance, you can follow the instructions in the [CONTRIBUTING](https://wsh032.github.io/fastapi-proxy-lib/CONTRIBUTING/CONTRIBUTING/#documentation).
